### ISSUE TITLE
Invalid sequence sent to UnmarshalMsgpack during Decode

### ISSUE THREAD

**Issue by lestrrat** [vmihailenco/msgpack#129]
While trying to test your fix for #127, I'm getting byte sequences that are slightly off.
The following patch, when run, results in a panic.
Upon closer inspection, this seems to be happening:

MarshalMsgpack produces byte sequence `[]byte{0x7, 0x5b, 0xcd, 0x15, 0x0, 0x0, 0x0, 0x7b}`, but the value that we get via `Decode` is `[]byte{0xd7, 0x0, 0x7, 0x5b, 0xcd, 0x15, 0x0, 0x0, 0x0}`. The msgpack type byte and the null padding seems to be included in the byte sequence passed to UnmarshalMsgpack

```diff
diff --git a/example_registerExt_test.go b/example_registerExt_test.go
index 9a2681d..a0b26cd 100644
--- a/example_registerExt_test.go
+++ b/example_registerExt_test.go
@@ -1,6 +1,7 @@
 package msgpack_test
 
 import (
+       "bytes"
        "encoding/binary"
        "fmt"
        "time"
@@ -16,6 +17,10 @@ type EventTime struct {
 var _ msgpack.Marshaler = (*EventTime)(nil)
 var _ msgpack.Unmarshaler = (*EventTime)(nil)
 
+func init() {
+       msgpack.RegisterExt(0, &EventTime{})
+}
+
 func (tm *EventTime) MarshalMsgpack() ([]byte, error) {
        b := make([]byte, 8)
        binary.BigEndian.PutUint32(b, uint32(tm.Unix()))
@@ -33,9 +38,7 @@ func (tm *EventTime) UnmarshalMsgpack(b []byte) error {
        return nil
 }
 
-func ExampleRegisterExt() {
-       msgpack.RegisterExt(0, &EventTime{})
-
+func ExampleRegisterExtMarshalUnmarshal() {
        b, err := msgpack.Marshal(&EventTime{time.Unix(123456789, 123)})
        if err != nil {
                panic(err)
@@ -49,3 +52,19 @@ func ExampleRegisterExt() {
        fmt.Println(v.(EventTime).UTC())
        // Output: 1973-11-29 21:33:09.000000123 +0000 UTC
 }
+
+func ExampleRegisterExtEncodeDecode() {
+       var buf bytes.Buffer
+       var t1 = EventTime{time.Unix(123456789, 123)}
+       if err := msgpack.NewEncoder(&buf).Encode(&t1); err != nil {
+               panic(err)
+       }
+
+       var t2 EventTime
+       if err := msgpack.NewDecoder(&buf).Decode(&t2); err != nil {
+               panic(err)
+       }
+
+       fmt.Println(t2.UTC())
+       // Output: 1973-11-29 21:33:09.000000123 +0000 UTC
+}
```

```
$ go test -run=Register -v
=== RUN   TestRegisterExtPanic
--- PASS: TestRegisterExtPanic (0.00s)
=== RUN   ExampleRegisterExtMarshalUnmarshal
--- PASS: ExampleRegisterExtMarshalUnmarshal (0.00s)
=== RUN   ExampleRegisterExtEncodeDecode
--- FAIL: ExampleRegisterExtEncodeDecode (0.00s)
panic: invalid data length: got 9, wanted 8 [recovered]
	panic: invalid data length: got 9, wanted 8

goroutine 1 [running]:
testing.runExample.func2(0xed099305b, 0xc436c5e8ca, 0x13941e0, 0xc42000c1e8, 0xc42000c018, 0xc42006ab40, 0x1274f1c, 0x1e, 0x127bad0, 0x1277d79, ...)
	/Users/daisuke/.gvm/gos/go1.8/src/testing/example.go:117 +0x359
panic(0x122ace0, 0xc420017500)
	/Users/daisuke/.gvm/gos/go1.8/src/runtime/panic.go:489 +0x2cf
gopkg.in/vmihailenco/msgpack%2ev2_test.ExampleRegisterExtEncodeDecode()
	/Users/daisuke/dev/src/gopkg.in/vmihailenco/msgpack.v2/example_registerExt_test.go:65 +0x4c2
testing.runExample(0x1274f1c, 0x1e, 0x127bad0, 0x1277d79, 0x28, 0x0, 0x0)
	/Users/daisuke/.gvm/gos/go1.8/src/testing/example.go:122 +0x204
testing.runExamples(0xc42003bed8, 0x138f5a0, 0xb, 0xb, 0x1230101)
	/Users/daisuke/.gvm/gos/go1.8/src/testing/example.go:46 +0x167
testing.(*M).Run(0xc42003bf20, 0xc42003bf20)
	/Users/daisuke/.gvm/gos/go1.8/src/testing/testing.go:823 +0x182
main.main()
	gopkg.in/vmihailenco/msgpack.v2/_test/_testmain.go:138 +0xf7
exit status 2
FAIL	gopkg.in/vmihailenco/msgpack.v2	0.011s
```

**Comment by vmihailenco** [comment]
The problem here is that you decode into `EventTime`, not `interface{}`. The fix is rather complex and affects all types so it is unlikely I can do it today.

**Comment by lestrrat** [comment]
Ohhhh, so I actually should NOT specify a concrete type on `Decode`? Didn't think about that one. (and for the record and benefit of the readers of this issue, yes, it did work after changing t2 to interface{})

Thanks for confirming, and no hurries.

Personally, this seems like it should either be fixed to work with concrete types or an error should be returned if you pass anything other than an interface{}.

**Comment by vmihailenco** [comment]
Hopefully fixed in v2.9.1

**Comment by lestrrat** [comment]
I think it's fixed :D
